src/levanter/data/shardcache.py

```diff
@@ -0,0 +1,712 @@
+
+        # kick off the writer tasks
+        self.writer_tasks = []
+        shard_assignments = [self.shard_source.shard_names[i::num_writers] for i in range(num_writers)]
+        worker_fn = ray.remote(resources=processor.resources())(_shard_writer_task)
+        for shard_names in shard_assignments:
+            self.writer_tasks.append(worker_fn.remote(self, shard_names))
+
+        self.shard_states = {name: _ShardState() for name in self.shard_source.shard_names}
+        self.reader_promises = {}
+        self.global_chunks = []
+
+        self._exception = None
+        self._exception_source = None
+
+    def finished_chunk(self, shard_name: str, chunk: ChunkMetadata):
+        """Called by a shard writer when it has finished writing a chunk. This will update the global ordering
+        of chunks and notify any readers that are waiting for this chunk."""
+        logger.debug(f"Finished writing chunk {chunk} for shard {shard_name}")
+
+        assert not self.shard_states[shard_name].finished
+        self.shard_states[shard_name].buffered_chunks.append(chunk)
+
+        # if we have a chunk for every incomplete shard, we can add them to the global ordering
+        self._check_chunks_ready()
+
+    def finished_shard(self, shard_name: str):
+        """Called by a shard writer when it has finished writing a shard. This will update the global ordering
+        of chunks and notify any readers that are waiting for this chunk."""
+        logger.debug(f"Finished writing shard {shard_name}")
+        self.shard_states[shard_name].finished = True
+        self._check_chunks_ready()
+
+    def shard_exception(self, shard_name: str, exc: _ExcInfo):
+        """Called by a shard writer when it has encountered an exception. This will notify any readers that
+        are waiting for this chunk."""
+        logger.error(f"Shard {shard_name} encountered an exception: {exc}", exc_info=_restore_exc_info(exc))
+        # only log the first exception for now
+        if self._exception is not None:
+            self._exception = exc
+            self._exception_source = shard_name
+
+            self._propagate_exceptions()
+
+    def other_exception(self, exc: _ExcInfo):
+        """Called by a shard writer when it has encountered an exception. This will notify any readers that
+        are waiting for this chunk."""
+        logger.error(f"Writing encountered an exception: {exc}", exc_info=_restore_exc_info(exc))
+        if self._exception is not None:
+            self._exception = exc
+            self._propagate_exceptions()
+
+    def _propagate_exceptions(self):
+        # send exceptions to all blocked readers, terminate all writers
+        for future in self.reader_promises.values():
+            future.set_exception(self._exception[0].with_traceback(self._exception[1].as_traceback()))
+        for task in self.writer_tasks:
+            ray.cancel(task)
+
+    async def get_chunk(self, global_index: int) -> Optional[ChunkMetadata]:
```

Comment:
Whoa, I didn't know that [async works with Ray actors](https://docs.ray.io/en/master/ray-core/actors/async_api.html#asyncio-for-actors). That's pretty cool.